### ISSUE TITLE
Move root assets to some public directory

### ISSUE THREAD

**Issue by bloatware** [textpattern/textpattern#1440]
### Is your feature request related to a problem?

Partially

### What is the feature?

We currently have (by default) three writeable directories in site's root: `files`, `images` and `themes`. This makes them unusable as section names in clean permlink modes. I suggest we find a way to move them to some `public` directory, transferring the upcoming `plugins` there too.

**Comment by Bloke** [comment]
If we can do it cleanly, sure. Just for new installs, right?

**Comment by bloatware** [comment]
Yes, changing the location of existing assets looks problematic. I'd start by moving `plugins` there, since it does not look clean to keep a writeable directory inside `textpattern` folder.

**Comment by Bloke** [comment]
I like this.

**Comment by jools-r** [comment]
`config.php` is another candidate. The multisite setup does this to a degree with separate `public`, `private` (config, templates (in the pre-themes days), maybe then also plugins etc.), and `admin` directories.

**Comment by bloatware** [comment]
Yes, everything 'custom' should be taken out of `textpattern` and go to a common directory. Easier for txp updates too. We also probably have `tmp` and `setup` to move?

**Comment by Bloke** [comment]
I'd forgotten about this. Is this doable in 4.9? Given that we're reinventing image handling and plugin handling, it might be a good time to bite this particular bullet and shuffle things around.

Not entirely sure how we'd manage it with upgrades since:

* images/themes are relative prefs
* files/tmp/plugin_cache is an absolute file path pref
* plugin_path is a constant
* setup is hard-coded.

Open to ideas!

**Comment by petecooper** [comment]
Marker to sift https://github.com/textpattern/textpattern/issues/1647 for useful discussion, especially regarding multisite capability.

**Comment by Bloke** [comment]
Okay so let's say we can do this somehow. How about this for a  layout:

- .htaccess
- css.php (can we get rid of this? Please?!)
- index.php
- /public/
  - images/
  - files/
  - themes/
- /shared/
  - tmp/
  - plugins/
  - config.php
  - config-dist.php
  - checksums.txt (?)
- /sites/
- /textpattern/
  - admin-themes/
  - include/
  - index.php
  - lang/
  - lib/
  - mode.ini
  - publish/
  - publish.php
  - setup/
  - textpattern.js
  - update/
  - vendors/

Essentially, for least upheaval, the textpattern directory stays intact, except for moving /plugins and /tmp out. a new top level /public directory houses all publicly-facing assets such as images, files and themes. I considered /plugins in here too but they're not always 'public' per se. To that end, I decided on a 'shared' directory name. For starters this more descriptive as it's for content shred between admin and public areas. So, plugins could be either. The tmp directory is a shared resource. And the config file.

The checksums.txt file could be moved. Or it could stay where it is in /textpattern. For added security, it could even become checksums.php and be converted to an array by our toolbelt script when regenerated. Or checksums.json. Removing the .txt extension and making it something we disallow being served, then it's one less place to leak the Txp version info.

I considered moving /setup to the top level but that has repercussions. Namely, we use the location of the setup directory to determine the admin-side URL and path (one directory hop higher), so it makes sense to leave that intact, rather than guess.

So that leaves /rpc. Where do we put that? Leave it where it is?

**Comment by petecooper** [comment]
>/shared/

Might be worth considering a prefix here (think `wp-`), even if it's CMS-agnostic and we go down the white labelling. For example: `cms-shared` or `cms-internal`…something less common for a domain subdir, essentially.

**Comment by Bloke** [comment]
Good idea, re: cheap "namespacing". Makes a lot of sense. I briefly considered `txp-shared` but `cms-shared` is better, since any bots crawling the directory get to associate the term with the project. I like your style.

I suspect 'textpattern' is fine but do we need to prefix 'public' too?

**Comment by Bloke** [comment]
Sure. The only reason I didn't rename textpattern was to make it easier for us to write the code. Without that, upgraders (who will keep /textpattern forevermore unless one day we force it to be changed on upgrade) need separate code or we need lots of "if version < a.b.c then use textpattern else use cms-admin unless upgrade then stick with textpattern" code, which slows everything down and hinders long-term maintenance.

Plus, people who upgrade will get parallel directories:

* /textpattern
* /cms-admin

which makes everything even more tricky. We could maybe get round it with redirects but if we do it at the htaccess level then Nginx have to do it by hand and not everyone replaces their .htaccess on upgrade because of heavy customisations (think: colak and his 3 million lines of bot filtering IP addresses).

It makes upgrades more challenging. I mean, we'll already have /images and /files in place and a parallel directory /cms-public/images and /cms-public/files etc. It is going to create a bit of a mess if we don't manage it well. While I don't like the idea of renaming directories and paths on upgrade, it does make everything simpler.

So, dunno. Part of me wants to make it less upheaval and keep code overhead low. Other parts of me want to just do it and force everyone to use the new structure - upgrades and new installs - somehow. Just don't know how to do it without killing a tonne of plugins (e.g. smd_thumbnail, smd_remote_file, any admin theme plugins or even themes themselves). It's all... messy.

**Comment by Bloke** [comment]
As part of this move, I'd like to consider images: specifically load balancing.

We can (I _think_) fairly easily modify our `imagesrcurl()` function to check the /images/id.ext location for a match via `is_readable()`. And if it fails, try our new /public/images/{some-load-balancing-dir}/{img-size}/id.ext second. And if that fails, 404. That function will thus return the true path of each image.

New installs will use the latter out of the gate. Upgraders will hit the former location. But if you choose to edit the file using the whizzy 4.9 Images interface, on saving, it makes sense to relocate it to the new location. The only time that will affect anybody is if a plugin _doesn't_ use `imagesrcurl()` and references images directly. Nothing to stop people still linking directly to named site assets, but anything stored in the database will potentially move.

So. Load balancing. Lots of options. While date-based makes a lot of sense, it does come with an overhead: you need to look up the file in the database to get its date and thus its location. There's also the minor annoyance that if you replace an image, we need to either move the image on save to a new date location, or leave it where it is and put up with the fact that the file's datestamp doesn't match its location.

If, however, we're sticking with image IDs as filenames (and there are a lot of programmatic advantages to keeping this convention, despite what people think about its SEO impact) then another option open to us is this structure:

```
/path/to/public/images/1/{img-size}/1.ext
/path/to/public/images/1/{img-size}/10.ext
/path/to/public/images/1/{img-size}/11.ext
/path/to/public/images/1/{img-size}/12.ext
...
/path/to/public/images/1/{img-size}/100.ext
/path/to/public/images/1/{img-size}/101.ext
...
/path/to/public/images/2/{img-size}/2.ext
/path/to/public/images/2/{img-size}/20.ext
/path/to/public/images/2/{img-size}/21.ext
/path/to/public/images/2/{img-size}/22.ext
...
/path/to/public/images/2/{img-size}/200.ext
...
```

Thus by simply looking at the ID of a file, we can immediately look it up in the filesystem by checking the first digit. From a load balancing viewpoint that's not perfect, but it's better than dumping them all in the same directory. It essentially divides our images-per-directory by 9.

Looking at a single number dir, It also means that if you use a few image sizes we get this:

```
/path/to/public/images/1/fullsize/1.ext
/path/to/public/images/1/fullsize/10.ext
/path/to/public/images/1/fullsize/11.ext
/path/to/public/images/1/fullsize/12.ext
/path/to/public/images/1/1280x960/1.ext
/path/to/public/images/1/1280x960/10.ext
/path/to/public/images/1/1280x960/11.ext
/path/to/public/images/1/1280x960/12.ext
/path/to/public/images/1/800x600/1.ext
/path/to/public/images/1/800x600/10.ext
/path/to/public/images/1/800x600/11.ext
/path/to/public/images/1/800x600/12.ext
/path/to/public/images/1/400x300/1.ext
/path/to/public/images/1/400x300/10.ext
/path/to/public/images/1/400x300/11.ext
/path/to/public/images/1/400x300/12.ext
```

So the number of subdirs inside the /1/ is only as big as the number of sizes you have. Not all images may be represented in all szes. You might change your mind over time and introduce other sizes for different resolutions or whatever. Doesn't matter too much. We can still find them all knowing the image ID first and then checking each size we need to display for a matching image.

We could swap the size and prefix. Not sure that's any better and it does mean that if you're using a file directory program to get to related files with the same number you need to jump up and back _two_ directories to get to each different file. Not a fan of that.

On a site like Keith's (towndock) site that has something on the order fo 26000 images, if he ever manages to go through the pain of editing each one over a number of years to move them all to the new structure, he's still looking at 3000 files per directory in each size. 3K in one directory is way better than 26k in one directory now, but can we do better?

Yes. Sort of. We could take the image ID and run it through a simple one- or two-way function. md5, xor, crc, anything. Then use maybe the first two, or three, or four digits of the result as the directory name, followed by the size. Not as immediate from a human perspective but still easily calculable to find an image location from its ID and desired resolution. I'd be tempted to put the full image path in the UI, btw, along with the image files sizes for each image. Just for sanity.

Either of the above schemes have a side benefit. If you request image ID 42 at 768 width and it's _not_ in the file system, Image Intervention could grab the original, apply a resize transform to it and then cache it for us in the 768/42.jpg. First access slow. Subsequent access immediate. So for people who don't want to create their images up-front or don't want to define sizes and then later change their mind and deiced to server different size files have two choices:

1) Use the admin interface for each image and tell it to go and create thumbs for the new sizes. Fast access on first load for every site visitor. Hassle for admin.
2) Don't. Let the system auto-create the thumbs you've asked for in your srcset on demand. Slow access on first load for one site visitor (which may be the admin!) then fast for everyone else. Minimal hassle for admin.

In both cases, it's up to the admin to manually tidy up old sizes if they want to later. From the admin interface it'll be easy to see which images have which size by clicking on them to see all the thumbs defined, but if you want to remove all files of a certain dimension all you need is a file search (*nix `find` will do) to grab the dir names and pipe them through `rm -f`. And if you accidentally delete a load you want, no harm. They'll get auto-recreated on next access by the front end. As long as the /fullsize images remain intact on the server, everything else is regeneratable (minus filters/operations you may have applied - but we'll worry about that later).

This system also supports simple replacement of any file with a different extension. Txp won't care. We can maybe get it to look for a resource matching the extension of the original and if it doesn't find it, strip off the extension and see if it finds an image at that size with a different extension.

If the storage system is offloaded to the cloud, nobody cares. You could change your image directory pref to point to a remote location. As long as you copy up all the /fullsize images, everything else will get generated (provided Intervention is given write access - again, we'll worry about that later).

Two things we need to support this, beyond Image Intervention (which I have in a local branch and am testing):

1) A new pref for image _path_. This is where the new /path/to/public/images will be stored. We need this because we need to know where the new images are kept compared to the old ones so we can try both locations.

Also, Intervention does a better job if it's given the path vs the URL. There's more data available. Plus it means EXIF and IPTC data can be read: a boon for plugins. If you have no choice but to supply a URL (because you decide to use cloud storage) then tough. You'll take the hit on not being able to get the file sizes and additional data, so such plugins will be useless or will need to work around things.

2) A load balancing storage system to limit the number of files per directory to something sane, that can scale to 10K, 50K, 100K, 250K, 1 million images without being unwieldy to find related images in a set, or slow down FTP programs to read dir contents.

Thoughts welcome on all the above.

**Comment by Bloke** [comment]
One more possible solution if we have a really good load balancer (i.e. not the simple 1..9) I posited first: drop the subdirs entirely:

Thus images would be stored:

```
/path/to/public/images/b1c4/1_full.ext
/path/to/public/images/b1c4/1_1280x960.ext
/path/to/public/images/b1c4/1_800x600.ext
/path/to/public/images/b1c4/28_full.ext
/path/to/public/images/b1c4/28_1280x960.ext
/path/to/public/images/b1c4/28_800x600.ext
/path/to/public/images/b1c4/28_400x300.ext
/path/to/public/images/912e/11_full.ext
/path/to/public/images/912e/11_1280x960.ext
/path/to/public/images/912e/11_400x300.ext
...
```

In that case, maybe image 1 and image 28 hash to the same 4-digit directory (just a guess: no idea) while the number 11 hashes to a different directory.

That gives us potentially more top-level directories immediately under /images, but if we can control these somehow with a clever, simple algorithm to limit them to, say, 1000 then it should still provide a good compromise of number of files per dir and dirs. And the flatter structure means less jumping around from dir to dir when managing the same file.

Since images can potentially be auto-created on the fly, it's no hardship to tag the resolution onto the file id.

Is that better?

**Comment by jools-r** [comment]
> Pick a prefix, apply to everything. It works for Automattic, so we'll probably be just fine. Viz:

Just came across [Bedrock](https://roots.io/bedrock/), an alternative WP setup that employs a quite different base structure for a WP installation. Maybe it offers some helpful clues to the original question of moving config and plugins out of the web root.

**Comment by Bloke** [comment]
Hehe, yeah, I did go off on a slight tangent there, oops.

So essentially the jumpoff point in the web-accessible directory (with index.php in it) is pretty much just a tiny kickstarter that goes and grabs the first wp file to kick things off. And the wp-config.php file that WordPress expects to be in the public root contains no info, apart from an include to go and fetch the real one that is _outside_ docroot. Fairly neat.

That's easily doable, in theory. Tricky on upgrade because if we include a similar 'kickstart' config.php file in our core bundle, it'll overwrite the real one on upload. So if we're going to do this we'll need to change its name to txp-config.php and then get people to move their real one outside.

It also slightly complicates installations, because you can't just upload the bundle in one location. Well, okay, you can if you set your site URL to run in a subdir! I bet someone will just upload the entire bundle to:

/var/www/example.tld/html/{everything-here}

which defeats the purpose of moving sensitive stuff out of a web-accessible location:

/var/www/example.tld/textpattern/{txp-core-stuff}
/var/www/example.tld/txp-config/{txp-core-config-files}
/var/www/example.tld/html/{assets-and-kickstart-etc}
...

Certainly something to consider though. If we can crack the upgrade nugget so it's not too onerous, I'm all for moving stuff around and getting as much as we can out of docroot.
